### ISSUE TITLE
🦠 Model Request: Ames Mutagenicity

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#434]
### Model Title

Ames Mutagenicity

### Publication

Hello @RoseZhen2022!

As part of your Outreachy contribution, we have assigned you the dataset "Ames Mutagenicity" from the Therapeutics Data Commons to try and build a binary classification ML model.
Please copy the provided Google Colab template and use this issue to provide updates on the progress. We'll value not only being able to build the model but also interpreting its results.

**Comment by RoseZhen2022** [comment]
Thanks @GemmaTuron , I am working on it.

**Comment by RoseZhen2022** [comment]
Hi @GemmaTuron , sorry for the late reply. Here is my report of the model:
**1  What are we trying to predict?**  
Dataset Description: Mutagenicity means the ability of a drug to induce genetic alterations. Drugs that can cause damage to the DNA can result in cell death or other severe adverse effects. Nowadays, the most widely used assay for testing the mutagenicity of compounds is the Ames experiment which was invented by a professor named Bruce Ames. The Ames test is a short-term bacterial reverse mutation assay detecting a large number of compounds which can induce genetic damage and frameshift mutations. Specifically, The Ames test uses several strains of the bacterium that carry mutations in genes involved in histidine]synthesis. These strains are auxotrophic mutants, i.e. they require histidine for growth, but cannot produce it. The method tests the capability of the tested substance in creating mutations that result in a return to a "prototrophic" state, so that the cells can grow on a histidine-free medium.
In the model, the predict probability were transformed to binary output 0/1 using cut-off 0.5. If the output is 1, it means the drug has more potential to be mutagenic.

**2. Is this a classification or a regression problem?** 
This is a classification to predict whether it is mutagenic (1) or not mutagenic (0).

**3.  How many datapoints do we have in total and in each set (train / validation / test)?** 
Totaly, there are 7278 samples in the dataset which is slightly different with the tdc website described as 7255 drugs. The dataset were splitted randomly into three sub-datasets. As a result, there are 5094 samples in train sub-dataset, 728 samples in validation sub-dataset, and 1456 samples the test sub-dataset. 

**4. How many actives and inactives do we have?** 
train_active: 2759,  train_inactive: 2335 ;
validation_active: 378, validation_inactive: 350 ;
test_active: 798, test_inactive: 658;

 **5. Plot show the distribution of actives/inactives in the train set:**

![image](https://user-images.githubusercontent.com/115380880/198505050-f0c4c8da-a135-4f04-be9e-3855c42ba3f4.png)



A example for inactive molecule (left) and a example for active molecule (right):

![image](https://user-images.githubusercontent.com/115380880/198505151-0fcaeb17-79c8-4021-97a9-3d19af4277c2.png)




**6.  The performance of model:**
·Using validation dataset, the performance of the model is 0.795951020534055. I also tried the test dataset, the performance is 0.7976590412200715, which is similiar as before.

**7. What metrics are you using and why?**
Here I used confusion matrix, a  kind of  contingency table, to show the data. It has two dimensions including prediction result as well as true result. By using this matrix, the performance of the model can be clearly and intuitively visualized. The first row of the matrix contains true negative (TP) and false positive (FP) samples. The second row contains false negative(FN) and (TP) samples. 
confusion matrix for the validation set:
![image](https://user-images.githubusercontent.com/115380880/198504963-6929e506-2484-458e-a4c5-1a0bc9f804e5.png)


In the figure showed above, we can calculate true positive rate (TPR) and false positive rate (FPR) as followes: 
TPR = TP/(TP + FN) = 338/(338+79) = 0.81
FPR = FP/(TN + FP) = 68/(243+68) = 0.22
Also, the precision (how many positives are actually positive) and recall (how many positives are we able to identify) can be calculated as follows:
precision = TP/(TP+FP) = 338/(338+68) = 0.83
recall = TP/(TP+FN)= 338/(338+79) = 0.81

confusion matrix for the test set:
![image](https://user-images.githubusercontent.com/115380880/198504909-9f647c60-0337-4d20-9ac0-4b0bcd7f063c.png)



TPR = TP/(TP + FN) = 640/(640+158) = 0.80
FPR = FP/(TN + FP) = 136/(136+522) = 0.21
precision = TP/(TP+FP) = 640/(640+136) = 0.82
recall = TP/(TP+FN)= 640/(640+158) = 0.80

**8. How do you interpret the ROC Curve you got?**
A ROC curve is constructed by plotting the true positive rate (TPR) against the false positive rate (FPR). The TPR is the proportion of observations that were correctly predicted to be positive out of all positive observations (TP/(TP + FN)). Similarly, the FPR is the proportion of observations that are incorrectly predicted to be positive out of all negative observations (FP/(TN + FP)). 
If a model gives the curve closer to the top-left corner, it means that it has a better performance. The “AUC” showed in the picture means the area under curve. Here the AUC on both the validation and test sets are 0.80, which means the model performance is good.
ROC curve for the validation set:

![image](https://user-images.githubusercontent.com/115380880/198504844-72a5ef31-32f9-4a91-aa78-a77925f72c33.png)

ROC curve for the test set:
![image](https://user-images.githubusercontent.com/115380880/198504871-c52cea33-62bf-4788-a1c5-10fab932c37f.png)

**Comment by RoseZhen2022** [comment]
And here is the link of my colab: https://colab.research.google.com/drive/1AGapnmCn-RSiHyqUZegErxv0o-qBPK2o?usp=sharing
I have learned a lot through this project. But I am not sure how to improve the performance of the model. Could you please give me some suggestions? Thank you very much!

**Comment by GemmaTuron** [comment]
Hi @RoseZhen2022 !

Good job! You can try to increase model training time for example

**Comment by RoseZhen2022** [comment]
Hi @GemmaTuron , I tried to increase the training time to 10 min, the performance of the model was improved to 0.84 on validation dataset. Here is the new ROC curve:
![image](https://user-images.githubusercontent.com/115380880/198827730-16239d9a-0780-4703-9695-f9bc75407c2d.png)

Thanks a lot!

**Comment by GemmaTuron** [comment]
Hi @RoseZhen2022 !

Thanks for the good work! I'll mark this as completed and you can go onto linking it to your Outreachy contributions and start working on your final application.
